### ISSUE TITLE
Trouble with the NIFTI writer

### ISSUE THREAD

**Issue by m-pilia** [simeks/stk#12]
When trying out the Sobel filter, I noticed that the NIFTI writer does not handle vector images. This is pretty bad if someone tries to store the registration results in NIFTI format. I guess the memory layout is not consistent with what NIFTI expects, but there may be more, since the scalar range is also wrong. The direction matrix is also wrong, but I guess that is still a TODO.

The test program I used is this
```c++
#include <stk/filters/sobel.h>
#include <stk/io/io.h>

int main(void)
{
    stk::Volume vol = stk::read_volume("in.nrrd");
    auto res = stk::sobel(vol);
    stk::write_volume("./out.vtk", res);
    return 0;
}
```
I tried it on a volume of size `103 × 94 × 161` stored in a compressed NRRD file, adjusting the extension of the output as needed, and what I get is

NIFTI
![nii_small](https://user-images.githubusercontent.com/8300317/44956106-f909fa00-aebe-11e8-8737-3ae069f5b6d6.png)
NRRD
![nrrd_small](https://user-images.githubusercontent.com/8300317/44956109-f909fa00-aebe-11e8-9f94-9044f246cffe.png)
VTK
![vtk_small](https://user-images.githubusercontent.com/8300317/44956111-f9a29080-aebe-11e8-90a7-0f8aa597d136.png)

**Comment by simeks** [comment]
Ahh, yeah. This rings a bell. I think nifti wants to transpose the image data, having the component as the outer axis. 

Nifti keeps bringing me headaches so I've been thinking about just scrapping it as the only reason I added it was because of the compression (which NRRD also has). At the same time, I think nifti is valuable as it's used a lot in brain imaging.

**Comment by m-pilia** [comment]
Yeah, NIFTI is a mess, but I think many of its design choices are actually better than other formats, even if it makes everything less trivial. Having the component on the most outside index makes sense, in many cases it is more efficient to treat channels as separate, contiguous arrays, especially for the sake of better vectorisation (it's like the famous rule of thumb "better a struct of arrays than an array of structs"). That's the design I also chose for disptools, and I also have this trouble of transposing stuff forth and back. I think the custom of having the channel in the fastest index was pushed mostly by ITK, because it makes easier to write generic code templated over the voxel type.

Even though NIFTI is messy, it so widespread we should probably have a minimal support for it. I opened this issue as a reminder, instead of just fixing the code, because I am working on other stuff these days, but I will probably have time to work on this within a few days.

All this is the very reason I usually prefer to not mess with IO natively. I have this sick idea, we could write an optional component that depends on ITK to read and write stuff. We let ITK read/write stuff, and then we move the data from/to a STK object. All we need is a simple adapter to convert between STK and ITK objects (which could turn to be useful anyway). This way we get complete support for a ton of formats for free, we would probably never have time to implement and maintain all of that ourselves. Of course ITK is a huge dependency, so I though we could keep the actual IO system as the native, default choice, and have this extra module turned off from the default build (that should always keep building without any ITK dependency).
